lib/spack/spack/stage.py

```diff
@@ -498,7 +498,8 @@ def cache_mirror(self, stats):
         absolute_storage_path = os.path.join(
             dst_root, self.mirror_paths.storage_path)
 
-        if os.path.exists(absolute_storage_path):
+        if (os.path.exists(absolute_storage_path) and
```

Comment:
If the storage path exists and the cosmetic path doesn't yet exist, then generally the only requirement is that we need to create the symlink. As is, this will re-download a file just to recreate the symlink. I realize that I didn't make this easy to accommodate but I'm wondering whether either (a) we can figure out how to just recreate the symlink here or (b) this can be moved to another PR.

Reply:
The fetch step happens before this, and regardless of whether these paths exist.
If we remove this it means mirrors won't fix the cosmetic paths unless we go in and delete the files at the storage paths. That's especially problematic in this case, as this whole issue means that we've lost the links to many of those storage paths in the first place.

Reply:
There's a call to `fetch` 5 lines down from here, that's where I intend to do the fetch; I intend that if the `storage_path` is already stored that we skip the fetch. Where do you observe that the fetch occurs before the call to `cache_mirror` (it is a mistake on my part that a fetch is occurring before this, is why I ask).

Reply:
Line 508 in mirror.py calls patch.fetch, which generates a stage just for the patch url and fetches it (on line 214 in patch.py).

Reply:
Also, I can't believe I missed that fetch in cache_mirror. It was, as you said, right there.

Reply:
Ah ok so there's a difference in how patches are handled vs. other resources. I think the call to `patch.fetch` in mirror can actually be removed, since that will be handled in the `cache_mirror` call. I think the easiest way to handle this would be to separate out the symlink creation in `MirrorCache` into a separate function, and have `Stage.cache_mirror` invoke this function regardless of whether the `storage_path` already exists. If you like that could be handled in a separate PR (by me or you, according to your preference); here works too.

Reply:
I did the separation in this merge request. I left the `patch.fetch` in mirror alone though.